### PR TITLE
feat(signin): send `marketingOptIn` parameter to server

### LINKED ISSUE #5195: Send marketingOptIn parameter when creating an account

### PR DESCRIPTION
**Pull request by seanmonstar**
Closes #5195

### CODE REVIEW

**Comment by shane-tomlinson** [comment]
@seanmonstar - this looks great! A couple of small requests so that we Do The Right Thing and ensure that we preserve flow event metrics semantics:

* [ ] Move [needsOptedInToMarketingEmail from PERSISTENT](https://github.com/mozilla/fxa-content-server/pull/5212/files#diff-046db0ab09009f68f513da9058e5be2eR36) to [DEFAULTS](https://github.com/mozilla/fxa-content-server/pull/5212/files#diff-046db0ab09009f68f513da9058e5be2eR57) so that it's no longer stored. If the value is stored, a 2nd attempt will be made to register on verification, and our flow event counts will be skewed.

* [ ] Lines R550:R553 triggers a flow event when the user subscribes to the newsletter. We'll want to ensure that logic is also called in `signUp` so we continue to track the # of users that opt-in. A `flow.initialize` will already have been triggered for the sign-up flow, so really all you need is R551:R553.

* After a few trains, when we can be reasonably sure most people have verified, we should be able to remove [the logic that subscribes via verifySignUp](https://github.com/mozilla/fxa-content-server/pull/5212/files#diff-046db0ab09009f68f513da9058e5be2eL544) too.

**Comment by shane-tomlinson** [comment]
Two things @seanmonstar informed me of that I didn't consider.

* The opt-in flag is stored for 24 hours in memcached. If the user verifies > 24 hours later, then the flag will not be in memcached and we'll want the fallback.
* flow-events are often keyed by the flow-id and event name. If a given flow-id can only have a single instance of a metric, we don't have to worry about duplicate entries. @philbooth - can you verify?

**Comment by philbooth** [comment]
> flow-events are often keyed by the flow-id and event name. If a given flow-id can only have a single instance of a metric, we don't have to worry about duplicate entries.

I disagree with the premise of this statement but not the conclusion.

Sometimes we key into the flow data by uid, so multiple events per metric can happen that way. And even when keying by flow id, it's still possible for one flow to have multiple instances of the newsletter subscription/unsubscription events. So duplicate entries can occur.

However, we also have the timestamp for each event, so we know whichever happened last and which events duplicated a pre-existing state. As long as we write the queries to take that into account (as far as I know, there aren't any queries of these events yet), we're golden.

**Comment by shane-tomlinson** [comment]
> I disagree with the premise of this statement but not the conclusion.

The explanation makes a lot of sense, thanks for clarifying our misconceptions!

**Comment by shane-tomlinson** [comment]
@seanmonstar - are you thinking of re-working this PR to use the resume token?

**Comment by seanmonstar** [comment]
Yea, gotta get some code into the auth server and fxa-js-client first...

**Comment by seanmonstar** [comment]
@shane-tomlinson something more like this? It still needs to be updated with a new fxa-js-client (waiting on PR there too).

File: app/scripts/models/resume-token.js [hunk 1]
```diff
@@ -19,6 +19,7 @@ define(function (require, exports, module) {
       entrypoint: undefined,
       flowBegin: undefined,
       flowId: undefined,
+      marketingOptIn: undefined,
```
Comment (shane-tomlinson) [inline app/scripts/models/resume-token.js]: This needs to be `needsOptedInToMarketingEmail` I think.

**Comment by shane-tomlinson** [comment]
> @shane-tomlinson something more like this? It still needs to be updated with a new fxa-js-client (waiting on PR there too).

YES! This should work well @seanmonstar! 

With this approach [needsOptedInToMarketingEmail](https://github.com/mozilla/fxa-content-server/pull/5212/files#diff-046db0ab09009f68f513da9058e5be2eR34) can safely be removed since the resume token never expires, though let's wait a train to ensure we have rollbacks covered. If `needsOptedInToLocalStorage` is not written to localStorage and then we roll back, we won't know to opt-in users. 

One thing to test is whether train upgrades work as expected. For example, in train-90 `needsOptedInToMarketingEmail` will be written to localStorage. If the user waits until train-92 to verify, when we are sure `needsOptedInToMarketingEmail` is no longer written to localStorage, `needsOptedInToMarketingEmail` should still be loaded from localStorage.

**Comment by seanmonstar** [comment]
@shane-tomlinson cool, I've added a comment to the field so we don't forget about that.

Also needs another fxa-js-client release...

**Comment by seanmonstar** [comment]
@shane-tomlinson r?

**Pull request merged**